teps/0010-optional-workspaces.md

```diff
@@ -0,0 +1,367 @@
+have no bearing on subsequent Tasks in a Pipeline.
+
+A Task can also conceivably skip some portion of work if the user has elected
+not to receive the generated data. For example, a Task may not need to compute
+the diff of its modifications or generate a manifest of changed files if the
+user has not provided a location for that data to be received.
+
+### Goals
+
+1. Allow Tasks to declare optional support for file-based credentials.
+2. Enable Tasks to offer support of Tekton's multiple competing credential mechanisms.
+3. Allow Tasks to optionally receive storage with some data to use during execution.
+4. Allow Tasks to optionally receive storage on to which the Task will place some output data.
+5. Enable Tasks to detect if any declared optional storage has been provided or not at runtime.
+
+### Non-Goals
+
+1. Declaring / binding specific file content on the optional storage.
+2. Validation of the file contents of the optional storage.
+3. Validation of "competing" filesystem fields used in a single TaskRun. An optional Workspace
+and an optional PipelineResource competing for the same location on the filesystem will be left
+up to the runtime to handle.
+
+## Requirements
+
+When optional storage is not provided by a TaskRun the Task must be able to
+detect this within the container / script.
+
+Tekton Variables should fail gracefully in the event that a workspace is
+not provided. Using a variable like `$(workspaces.foo.path)` should not
+cause the controller to abandon or fail a TaskRun when the optional workspace
+`foo` is omitted by it. Likewise the literal string `"$(workspaces.foo.path)"`
+should not end up being passed to the Task.
+
+When a Tekton deployment includes configuration for
+[Default TaskRun Workspaces](https://github.com/tektoncd/pipeline/blob/3a7a693429c605a67e30efa5b74675e403be79bb/config/config-defaults.yaml#L66-L70)
+a missing optional Workspace does not result in a Default TaskRun Workspace
+being injected in its place. The reason for this is that optional behaviour
+in a Task should not be switched on automatically by the Default TaskRun
+Workspace mechanism.
+
+## Proposal
+
+Allow Workspaces to be declared "optional" in Tasks. An optional Workspace is
+not required for successful execution of the Task but can be provided by a TaskRun
+to configure some optional behaviour, like using a specific kind of credential.
+
+An optional Workspace provides the following variables:
+
+- `$(workspaces.<workspace-name>.bound)`: the literal string `"true"` or `"false"`
+reflecting whether the optional workspace has been bound or not.
+- `$(workspaces.<workspace-name>.path)`: the path to the workspace if it has been
+bound _or_ an empty string if it has not been.
+
+A required Workspace today already provides a `path` variable. Required Workspaces
+will be updated to also expose the following variable:
+
+- `$(workspaces.<workspace-name>.bound)`: the literal string `"true"`. A required
```

Comment:
I'm personally in favor of the explicit `bound` variable but it mostly comes down to readability / explicitness for me.

Relying on the `path` being an empty string as a signal of bound-ness feels to me like a kind of institutional knowledge: if you "know" that that's what it signifies then you'll have no problem reading a Task, but if you don't have that knowledge it will just ... look a bit odd.  You'll have to go look at some docs and come back and try to figure out the meaning of this slightly idiosyncratic bit of Tekton's syntax.  On the other hand, the language of workspaces already includes "bound" - it's in the type itself (`WorkspaceBinding`) and repeated in loads of places in the `workspaces.md` doc already.

In some ways I wish the fields were more explicitly named too: Tasks could include a `workspaceDeclarations` section and Runs could include a `workspaceBindings` section.  Then the language would tie even more closely together.  But I think I've probably missed the boat on that one.

Reply:
one more counter point for bound:

> if you "know" that that's what it signifies then you'll have no problem reading a Task, but if you don't have that knowledge it will just ... look a bit odd

There will be nothing to _force_ folks to use _bound_ when they can get the same information from the path being the empty string, so having 2 ways of representing this information might be make it even harder to read Tasks using this feature?

That's the last thing I'll say about it tho, happy to defer to your preference having said my piece :D